Add TS+ Post-Mortem blog post (#1173)
Effect-TS/website: content/src/content/docs/blog/ts-plus-postmortem.mdx

5. **Operator Overloading**: Enable custom operators for domain-specific libraries, making mathematical and functional code more expressive
6. **Functional Pipe Operator**: Implement a native pipe operator for cleaner data transformation pipelines

The project was born from our belief that language tooling could be pushed further to create magical developer experiences—where types don't just prevent errors, but actively generate useful code, and where APIs feel natural and discoverable.

## What Was Achieved

Despite being an experimental fork, TS+ successfully demonstrated several groundbreaking capabilities that pushed the boundaries of what's possible with TypeScript:

**Fluent, Tree-Shakable APIs**: We proved that extending types with methods while maintaining optimal bundle sizes was not only possible but practical:

```typescript
// Traditional approach
pipe(
  IO.succeed(0),
  IO.map(n => n + 1),
  IO.flatMap(r => IO.succeedWith(() => console.log(`result: ${r}`)))
)

// TS+ fluent API
const program = IO.succeed(0)
  .map(n => n + 1)
  .flatMap(r => 
    IO.succeedWith(() => {
      console.log(`result: ${r}`);
    })

>   );

Reviewer: It might be worthwhile to have a comment here that indicates that this code, when compiled, outputs code similar to the above. On first read, while I know this is the case, I guarantee another reader will not.

program.run();
```

**Runtime Type Derivation**: Successfully implemented automatic generation of runtime code from type definitions:

```typescript
// Define once, derive everything
interface Person {
  name: string;
  age: Option<number>;
  friends: Chunk<Person>;
}

// Automatically generated codecs, decoders, guards
export const PersonEncoder: Encoder<Person> = Derive();
export const PersonDecoder: Decoder<Person> = Derive();

> export const PersonGuard: Guard<Person> = Derive();

Reviewer: Again, I think it is worth "highlighting" the `Derive` call here somehow to the reader's attention to the fact that all the code will be generated. Perhaps even showing a small snippet of the code that might be generated (even if it's not accurate).

I think showing examples of the code generation will help the reader to understand how impactful this is.

// Usage in practice
const encoded = PersonEncoder.encodeJSON({
  name: "Mike",
  age: Option.some(30),
  friends: Chunk()
});
```

**Extensible Method Definitions**: Created a system where developers could add methods to existing types without modifying original classes:

```typescript
/**
 * @tsplus type IO
 */
export class IO<A> {
  constructor(readonly io: () => A) {}

> }

Reviewer: I would suggest giving an example below the bullet so the reader knows what they are looking at.

Something like:
> The `@tsplus type IO` directive registers the `IO` type with TS+ under the name `IO`. Once that is done, other directives can be used to augment the `IO` type. For example, `@tsplus fluent IO map` registers the `map` function as a fluent method on the `IO` type, allowing APIs to be chained together without attaching them to the same runtime object.

I can't remember if the compiler gave us errors when we mistyped these directives, but if it did might be worth also mentioning that here.

/**
 * @tsplus fluent IO map
 */
export function map<A, B>(self: IO<A>, f: (a: A) => B): IO<B> {
  return new IO(() => f(self.io()));
}

/**
 * @tsplus fluent IO flatMap
 */
export function flatMap<A, B>(self: IO<A>, f: (a: A) => IO<B>): IO<B> {
  return new IO(() => f(self.io()).io());
}

> // Usage becomes natural - methods appear on the type:

Reviewer: ```suggestion
// Usage is intuitive - fluent methods are discoverable via autocomplete
```

const result = IO.succeed(10)
  .map(x => x * 2)
  .flatMap(x => IO.succeed(x + 5));
```

**Custom Operator Support**: Demonstrated that TypeScript could support domain-specific operators:

```typescript
/**
 * @tsplus fluent IO zip
 * @tsplus operator IO +
 */
export function zip<A, B>(self: IO<A>, b: IO<B>): IO<[A, B]> {
  return IO(() => [self.io(), b.io()]);
}

// Usage: combining IOs with natural operator syntax
const zipped = IO.succeed(0) + IO.succeed(1);
// Result: IO<[number, number]>
```

**Enhanced Do Syntax**: Made effect-based programming more intuitive by introducing a clean, imperative-style syntax for monadic computations:

```typescript

> // Traditional Effect code

Reviewer: ```suggestion
// Traditional Effect code (prior to the removal of the `_` adapter)
```

  const profile = yield* _(getProfile(user.id))
  const settings = yield* _(getSettings(user.id))
  return { user, profile, settings }
})

// TS+ do syntax - same logic, cleaner syntax
const effectExample = Do(($) => {
  const user = $(getUser(id));
  const profile = $(getProfile(user.id));
  const settings = $(getSettings(user.id));
  return { user, profile, settings };
});
```

**Global Import System**: Streamlined imports while maintaining tree-shaking:

```typescript
// Traditional approach: explicit imports in every file
import { Chunk } from "@tsplus/stdlib/collections/Chunk/definition";
import { Option } from "@tsplus/stdlib/data/Option/definition";
import { IO } from "./IO";

const processData = (items: Chunk<string>) => {
  return items.map(item => Option.some(item));
};

// TS+ global imports: define once in prelude.d.ts
/**
 * @tsplus global
 */
import { Chunk } from "@tsplus/stdlib/collections/Chunk/definition";

/**
 * @tsplus global
 */
import { Option } from "@tsplus/stdlib/data/Option/definition";

// Usage: no imports needed, types available globally
const processData = (items: Chunk<string>) => {
  return items.map(item => Option.some(item));

> };

Reviewer: Indicate this is in a separate file somehow.

We didn't approach TS+ as a casual experiment—we went all-in. For roughly a year, we committed significant resources and energy to prove that this approach could work at scale.

**Deep Technical Investment**: We spent countless hours in the depths of the TypeScript compiler codebase, understanding its intricate parsing, type-checking, and compilation phases. This wasn't surface-level modification; we were fundamentally extending the language's capabilities while maintaining backward compatibility.

**Complete Ecosystem Rewrite**: Perhaps most significantly, we rewrote the entire Effect ecosystem using TS+. This meant porting thousands of lines of carefully crafted functional programming abstractions, rebuilding our core libraries, and ensuring that all the complex type-level computations still worked correctly with our enhanced compiler.

**Production Adoption**: The experiment reached a point where we had production users depending on alpha releases. Real projects, real applications, real teams betting their development workflow on our experimental fork. This level of adoption validated that TS+ wasn't just a theoretical exercise—it was solving genuine developer pain points.

> The scale of our commitment demonstrated our conviction that language tooling could be dramatically improved. We weren't just building a proof-of-concept; we were building a complete alternative development experience.

Reviewer: ```suggestion
The scale of our commitment demonstrated our conviction that language tooling could be dramatically improved. We weren't just building a proof-of-concept; we were building a completely different development experience.
```
